File: current_session/986.py
"""
https://leetcode.com/problems/interval-list-intersections/
"""
class Solution(object):
    def intervalIntersection(self, A, B):
        """
        :type A: List[List[int]]
        :type B: List[List[int]]
        :rtype: List[List[int]]
        """
        # change A and B into series of 'start' and 'end' events
        events = []
        for a in A:
            events += (a[0], 'b'),
            events += (a[1], 'e'),
        for b in B:
            events += (b[0], 'b'),
            events += (b[1], 'e'),
        events.sort()

        # optimization: use 2 cursors and get rid of the above part
        res, cntOfIntvl, s = [], 0, None
        for i in range(len(events)):
            if events[i][1] == 'b':
                cntOfIntvl += 1
                if cntOfIntvl == 2:
                    s = events[i][0]
            else:
                cntOfIntvl -= 1
                if cntOfIntvl == 1:
                    res += [s, events[i][0]],

        # print res
        return res

print(Solution().intervalIntersection(A = [[0,2],[5,10],[13,23],[24,25]], B = [[1,5],[8,12],[15,24],[25,26]]))

File: current_session/c++/986.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
    vector<pair<int,int>> events;
    for (auto e:firstList) {
      events.push_back(pair<int,int>{e[0], 0});
      events.push_back(pair<int,int>{e[1], 1});
    }
    for (auto e:secondList) {
      events.push_back(pair<int,int>{e[0], 0});
      events.push_back(pair<int,int>{e[1], 1});
    }
    sort(events.begin(), events.end());
    int startCnt = 0;
    int startTime = INT_MAX;
    vector<vector<int>> res;
    for (auto x: events) {
      if (x.second%2 == 0) {
        startCnt += 1;
      } else {
        startCnt -= 1;
      }
      if (startCnt == 2) {
        startTime = x.first;
      } else if (startTime != INT_MAX) {
        res.push_back(vector<int>{startTime, x.first});
        startTime = INT_MAX;
      }
    }
    return res;
  }
};